### PR TITLE
SONARJAVA-1295 Add support for try catch finally in CFG

### PR DESCRIPTION
**Pull request by DidierBesset**

### CODE REVIEW

File: java-squid/src/main/java/org/sonar/java/cfg/CFG.java [hunk 1]
```diff
@@ -346,6 +371,39 @@ private void build(Tree tree) {
...
+          currentBlock.successors().add(catchBlock.block());
+        }
+      }
+      if (finallyBlock != null) {
+        currentBlock.successors().add(finallyBlock);
+      }
+    }
+  }
+
+  private boolean canRaise(Block block, SyntaxToken exception) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/cfg/CFG.java]: MAJOR Make "canRaise" a "static" method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2325)

File: java-squid/src/main/java/org/sonar/java/cfg/CFG.java [hunk 2]
```diff
@@ -767,8 +844,13 @@ private static String syntaxNodeToDebugString(Tree syntaxNode) {
       case INT_LITERAL:
         sb.append(' ').append(((LiteralTree) syntaxNode).token().text());
         break;
-      default:
-        //no need to debug other syntaxNodes
+      case METHOD_INVOCATION:
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/cfg/CFG.java]: MAJOR Reduce this switch case number of lines from 6 to at most 5, for example by extracting code into methods. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1151)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 3 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Add a default case to this switch.](https://github.com/SonarSource/sonar-java/blob/901329052ed86c7fc13d6291012cd319d69f0997/java-squid/src/main/java/org/sonar/java/cfg/CFG.java#L837) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ASwitchLastCaseIsDefaultCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 3 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Add a default case to this switch.](https://github.com/SonarSource/sonar-java/blob/901329052ed86c7fc13d6291012cd319d69f0997/java-squid/src/main/java/org/sonar/java/cfg/CFG.java#L837) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ASwitchLastCaseIsDefaultCheck)

File: java-squid/src/main/java/org/sonar/java/cfg/CFG.java [hunk 3]
```diff
@@ -346,6 +371,39 @@ private void build(Tree tree) {
...
+        }
+      }
+      if (finallyBlock != null) {
+        currentBlock.successors().add(finallyBlock);
+      }
+    }
+  }
+
+  private boolean canRaise(Block block, SyntaxToken exception) {
+    // FIXME this method should return true if the block can raise the supplied exception
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/cfg/CFG.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1134)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 5 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Add a default case to this switch.](https://github.com/SonarSource/sonar-java/blob/901329052ed86c7fc13d6291012cd319d69f0997/java-squid/src/main/java/org/sonar/java/cfg/CFG.java#L837) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ASwitchLastCaseIsDefaultCheck)
* MAJOR [Take the required action to fix the issue indicated by this comment.](https://github.com/SonarSource/sonar-java/blob/901329052ed86c7fc13d6291012cd319d69f0997/java-squid/src/main/java/org/sonar/java/cfg/CFG.java#L526) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1134)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 5 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Add a default case to this switch.](https://github.com/SonarSource/sonar-java/blob/901329052ed86c7fc13d6291012cd319d69f0997/java-squid/src/main/java/org/sonar/java/cfg/CFG.java#L837) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ASwitchLastCaseIsDefaultCheck)
* MAJOR [Take the required action to fix the issue indicated by this comment.](https://github.com/SonarSource/sonar-java/blob/901329052ed86c7fc13d6291012cd319d69f0997/java-squid/src/main/java/org/sonar/java/cfg/CFG.java#L526) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1134)

**Comment by benzonico** [comment]
Fixed by https://github.com/SonarSource/sonar-java/pull/531

**Pull request closed without merging**